torch/csrc/autograd/variable.cpp

```diff
@@ -21,91 +21,18 @@
 
 namespace torch {
 namespace autograd {
-Variable::Impl::Impl(at::Tensor data, std::unique_ptr<Variable::AutogradMeta> autograd_meta, bool requires_grad, Edge gradient_edge)
-    : TensorImpl(data.type_id(), data.dtype(), /*allocator=*/nullptr, /* is variable */ true),
-      data_(std::move(data)) {
-  autograd_meta->grad_fn_ = std::move(gradient_edge.function);
-  autograd_meta->requires_grad_ = false;
-  autograd_meta->is_view_ = false;
-  autograd_meta->output_nr_ = gradient_edge.input_nr;
-  autograd_meta->pyobj_ = nullptr;
+Variable::AutogradMeta::AutogradMeta(at::TensorImpl* self_impl, bool requires_grad, Edge gradient_edge) {
+  grad_fn_ = std::move(gradient_edge.function);
+  requires_grad_ = false;
+  is_view_ = false;
+  output_nr_ = gradient_edge.input_nr;
+  pyobj_ = nullptr;
 
   // set_requires_grad also checks error conditions.
-  autograd_meta->set_requires_grad(requires_grad, this);
+  set_requires_grad(requires_grad, self_impl);
```

Comment:
So, if I understand correctly, `set_requires_grad` only really needs the dtype, is that right? Might be worth cleaning up after this patch.

Reply:
Will do after this patch.